lib/datasource/github-tags/index.ts

```diff
@@ -98,6 +104,92 @@ export async function getDigest(
   return digest;
 }
 
+type GithubTag = {
+  name: string;
+  commit?: {
+    sha?: string;
+    url?: string;
+  };
+};
+
+type GithubCommit = {
+  commit?: {
+    author?: {
+      date?: string;
+    };
+  };
+};
+
+type GithubCommitTimestamps = Record<string, string | null>;
+
+async function getReleasesWithTimestamp(
+  repo: string,
+  tags: GithubTag[] | null
+): Promise<Release[]> | null {
+  const result: Release[] = [];
+
+  const cached =
+    (await renovateCache.get<GithubCommitTimestamps>(
+      cacheNamespace,
+      getCacheKey(repo, 'commit-timestamps')
+    )) || {};
+
+  let updateCache = false;
+
+  const queue = tags.map(tag => (): Promise<void> => {
+    const release: Release = {
+      version: tag.name,
+      gitRef: tag.name,
+    };
+
+    const commitHash = tag?.commit?.sha;
+
+    const finalize = (releaseTimestamp: string = null): void => {
+      cached[commitHash] = releaseTimestamp;
+      if (releaseTimestamp) {
+        release.releaseTimestamp = releaseTimestamp;
+      }
+      result.push(release);
+    };
+
+    const commitUrl = tag?.commit?.url;
+
+    if (commitUrl && commitHash) {
+      if (cached[commitHash] === undefined) {
+        updateCache = true;
+        return ghGot<GithubCommit>(commitUrl, {
+          paginate: true,
+        })
+          .then<void>(res => {
+            const releaseTimestamp = res.body.commit?.author?.date;
+            return finalize(releaseTimestamp);
+          })
+          .catch(err => {
+            logger.debug({ repo, err }, 'Error retrieving github commit');
+            finalize();
+          });
+      }
+      finalize(cached[commitHash]);
+    } else {
+      finalize();
+    }
+    return null;
+  });
+
+  await pAll(queue, { concurrency: 5 });
+
+  if (updateCache) {
+    await renovateCache.set<GithubCommitTimestamps>(
+      cacheNamespace,
+      getCacheKey(repo, 'commit-timestamps'),
+      cached,
+      2 * 24 * 60
```

Comment:
Is this correct? We are caching a mapping from *commit hash* to timestamp? i.e. something that maybe even never changes?

I wanted to make sure we're not caching *tag* to timestamp, which could change more frequently.

Reply:
Yes, it's caches by commit hash (see `setReleaseTimestamp()` above), but I just realized it can be never reset for frequently updated repos. Now should work fine.

Reply:
So as it currently stands, for a repo with no new tags we would re-query one per tag every two days?

Reply:
Exactly

Reply:
if we cache commithash to timestamp, this should never change, so we can cache forever?

Reply:
"forever" would bite us if we ever discover a problem, whereas a shorter time frame lets it self-heal at least. Maybe start with 2 days and see

Reply:
For now, there is a per-repo bulk cache (i.e. its values are `Record<string, string | null>`) which resets each 48 hours. With this configuration, infinite TTL means we won't be able obtain timestamps for fresh tags.

Another option is per-repo cache which updates it's value when new tags appear. In this case, for frequently updated repos it effectively will be stored forever. For most repos will be reset after 48 hours.

Ideally, I would store it for each hash separatedly with infinite TTL. But seems like we'll have too much pairs, whose keys are bigger than values?

So yes, I'm unsure which option is preferrable in terms of cache memory/IO tradeoff.

Reply:
I think we should cache commit hash to timestamp in long term cache (`global.renovateCache`) and store it for 48h as start. so other repos would benefit from cache. Later we can extend to cache time to 1w or more if everything works as expected.

Reply:
@zharinov it should work like this:

- Fetch a fresh list of tags and their gitrefs
- Fetch our cached gitrefs -> timestamp list
- If any we need are missing, then query them one by one
- Save the new collection to the cache over the top of the old cache

Using the above approach means you avoid this problem:
> infinite TTL means we won't be able obtain timestamps for fresh tags.